ramips: convert MT7621 wireless calibration EEPROM to NVMEM format

DragonBluep opened the pull request:
Notice:
1. Some MT7621 devices have complex dtsi dependencies, so they are not included.
2. All NIC models are judged based on the commit message or the makefile default packages.

|Package Name|NIC Model|EEPROM size||
|-|-|-|-|
|kmod-mt7603|MT7603|0x400|2 GHz only|
|kmod-mt76x0e|MT7610|0x200|same as MT7612|
|kmod-mt76x2|MT7602/MT7612|0x200||
|kmod-mt7663-firmware-ap|MT7613|0x4da8|same as MT7615|
|kmod-mt7615-firmware|MT7615|0x4da8||
|kmod-mt7915-firmware|MT7915|0xe00||

DragonBluep committed:
ramips: convert MT7610 and MT7612 EEPROM to NVMEM format for MT7621

This patch converts MT7610 and MT7612 WiFi calibration data to NVMEM
format. The EEPROM size is 0x200.

Signed-off-by: Shiji Yang <yangshiji66@qq.com>

DragonBluep committed:
ramips: convert MT7613 and MT7615 EEPROM to NVMEM format for MT7621

This patch converts MT7613 and MT7615 WiFi calibration data to NVMEM
format. The EEPROM size is 0x4da8.

Signed-off-by: Shiji Yang <yangshiji66@qq.com>

DragonBluep committed:
ramips: convert MT7915 EEPROM to NVMEM format

This patch converts MT7915 WiFi calibration data to NVMEM format. The
EEPROM size is 0xe00.

Signed-off-by: Shiji Yang <yangshiji66@qq.com>

DragonBluep committed:
ramips: convert MT7603 EEPROM to NVMEM format

This patch converts MT7603 WiFi calibration data to NVMEM format. The
EEPROM size is 0x400.

Signed-off-by: Shiji Yang <yangshiji66@qq.com>

neheb commented on target/linux/ramips/dts/mt7621_linksys_e7350.dts:
```diff
@@ -127,7 +134,8 @@
 	wifi@0,0 {
```
now I'm curious what happens if this section is removed. Does WiFi go away?

csharper2005 commented on target/linux/ramips/dts/mt7621_snr_snr-cpe-me1.dts:
```diff
@@ -144,13 +145,20 @@
+				compatible = "nvmem-cells";
 				label = "Factory";
 				reg = <0x40000 0x10000>;
-				read-only;
```
Why are we remove `read-only`?

raenye commented:
Does NVMEM format make it easier to patch EEPROM via DTS?
(instead of going through the `insmod mtr_rw i-want-a-brick` route)

DragonBluep commented:
> Does NVMEM format make it easier to patch EEPROM via DTS? (instead of going through the `insmod mtr_rw i-want-a-brick` route)

No, NVMEM means `Non Volatile Memory`. It can't be edited.

dangowrt commented:
> No, NVMEM means `Non Volatile Memory`. It can't be edited.

Sorry, but that's not entirely correct. Quoting Wikipedia: "Non-volatile random-access memory (NVRAM) is random-access memory that retains data without applied power." It can totally offer read-write access.

Using the nvmem framework in Linux on top of an MTD device results in read-only access simply because write access isn't implemented in the NVMEM MTD backend driver (because one can just use MTD API to write, for other NVMEM devices, e.g. AT24/AT25 I2C EEPROMs, write via NVMEM API *is* implemented).

tldr; This is unrelated. You will still need `mtd-rw` to remove the read-only flag from MTD partitions if you wish to modify their content. Nothing to do with whether the driver uses nvmem-cells or our previous mtd-eeprom hack to access the WiFi EEPROM data.

Ansuel reviewed (changes requested):
Man I think after this review, I'm ready for a run of Paper Please...

Generally this looks ok. I notice there are additional cleanup but they are ok and can stay considering the amount of work done. Just some minor tweak to do.

Also please do them but don't rebase this on top of main so I can check the changes by comparing the commit before and after. (if you are asking yes I checked all the 148 file)

DragonBluep committed:
ramips: correct NVMEM MAC address node name and label for Bolt Arion

The node name&label should match the address in the 'reg' property,
so it's better to change the incorrect offset to the 0x28.

Signed-off-by: Shiji Yang <yangshiji66@qq.com>

DragonBluep commented on target/linux/ramips/dts/mt7621_tplink_ec330-g5u-v1.dts:
```diff
@@ -225,25 +225,30 @@
 				#address-cells = <1>;
 				#size-cells = <1>;
 
+				eeprom_factory_8000: eeprom@8000 {
+					reg = <0x8000 0x4da8>;
+				};
+
+				eeprom_factory_14000: eeprom@14000 {
+					reg = <0x14000 0x4da8>;
+				};
```
It should be okay. I saw @rmilecki pushed a similar commit. https://github.com/openwrt/openwrt/commit/03b328480559a0183752f668dece2ff00d3bee28

DragonBluep committed:
ramips: fix EEPROM size for I-O DATA WN-DEAX1800GR

I-O DATA WN-DEAX1800GR uses MT7915 PCIe NIC. The correct EEPROM
size is 0xe00.

Fixes: ac68fbf526d ("ramips: add support for I-O DATA WN-DEAX1800GR")
Signed-off-by: Shiji Yang <yangshiji66@qq.com>

DragonBluep commented:
@Ansuel Hi! Thanks for your review. I've made the changes according to your suggestion. And when I checked if nesting eeprom cell into nvmem-layout was the correct way, I found the EEPROM size of the I-O DATA WN-DEAX1800GR was wrong, so I added another small commit to fix it. https://github.com/openwrt/openwrt/pull/13587/commits/6b74f1b708ed6c5f1709312300c51d854e94cf71

Ansuel commented on target/linux/ramips/dts/mt7621_bolt_arion.dts:
```diff
@@ -100,9 +100,24 @@
+				};
+
+				macaddr_factory_4000: macaddr@4000 {
+					reg = <0x28 0x6>;
```
Idea for this fix was to change this in a previous commit and then make the convert commit just move the node inside.
Since you have to touch this pr again, can you do this change?

DragonBluep replied:
You mean nesting nvmem stuff into partition nodes in the previous commit? It should be fine now.
https://github.com/openwrt/openwrt/pull/13587/commits/6f52c1429a29e0594df10d7df223366a41e29ff9

Ansuel replied:
No I mean that before all these commit, you create a single simple commit that just fix the wrong node name and tag.

And then in the conversion you move the node in the partition nodes as you did in the linked commit.

DragonBluep replied:
Oh, it seems that you didn't notice this small commit before, I've already done it. https://github.com/openwrt/openwrt/commit/001f05b220e3d2a0419026ce50332c8b9a33cc69
I'll force push it. Thanks again. It's really a hard job to review these changes.

Ansuel replied:
Oh it was there and I didn't notice! Sorry!!

raenye commented:
Please see issues #13700 and #13701 (mt7915).

neheb commented:
this is not in 23.05.0

raenye commented:
@neheb #13701 is indeed mt7622 so probably not related, but #13700 is snapshot.
